### PR TITLE
Refactor error handling in CustomDjangoCache

### LINKED ISSUE #11767: Possibility of unhandled exception during diskcache usage
## Observed behavior
A very odd behaviour whereby diskcache seems to lose track of its transaction id.

## Errors and logs
```
ERROR 2024-01-23 15:39:39,832 django.request Internal Server Error: /coach/api/notifications/
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kolibri/dist/diskcache/core.py", line 762, in _transact
    yield sql, filenames.append
  File "/usr/lib/python3/dist-packages/kolibri/dist/diskcache/core.py", line 827, in set
    rows = sql(
sqlite3.OperationalError: disk I/O error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/handlers/base.py", line 172, in _get_response
    resolver_match = resolver.resolve(request.path_info)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/urls/resolvers.py", line 366, in resolve
    sub_match = pattern.resolve(new_path)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/urls/resolvers.py", line 366, in resolve
    sub_match = pattern.resolve(new_path)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/urls/resolvers.py", line 366, in resolve
    sub_match = pattern.resolve(new_path)
  [Previous line repeated 1 more time]
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/urls/resolvers.py", line 361, in resolve
    match = self.regex.search(path)
  File "/usr/lib/python3/dist-packages/kolibri/core/device/translation.py", line 158, in regex
    device_language = get_device_language() or get_settings_language()
  File "/usr/lib/python3/dist-packages/kolibri/core/device/translation.py", line 26, in get_device_language
    if not device_provisioned():
  File "/usr/lib/python3/dist-packages/kolibri/core/device/utils.py", line 68, in device_provisioned
    return get_device_setting("is_provisioned")
  File "/usr/lib/python3/dist-packages/kolibri/core/device/utils.py", line 54, in get_device_setting
    device_settings = DeviceSettings.objects.get()
  File "/usr/lib/python3/dist-packages/kolibri/core/device/models.py", line 73, in get
    cache.set(DEVICE_SETTINGS_CACHE_KEY, model, 600)
  File "/usr/lib/python3/dist-packages/kolibri/deployment/default/custom_django_cache.py", line 76, in set
    return super(CustomDjangoCache, self).set(
  File "/usr/lib/python3/dist-packages/kolibri/dist/diskcache/djangocache.py", line 147, in set
    return self._cache.set(key, value, timeout, read, tag, retry)
  File "/usr/lib/python3/dist-packages/kolibri/dist/diskcache/fanout.py", line 97, in set
    return shard.set(key, value, expire, read, tag, retry)
  File "/usr/lib/python3/dist-packages/kolibri/dist/diskcache/core.py", line 842, in set
    return True
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/lib/python3/dist-packages/kolibri/dist/diskcache/core.py", line 765, in _transact
    assert self._txn_id == tid
AssertionError
```

## Expected behavior
Our CustomDjangoCache object should also handle AssertionErrors: https://github.com/learningequality/kolibri/blob/fc94f279c8affe2a5c69fc8f4c0879b9fb804b41/kolibri/deployment/default/custom_django_cache.py#L7

Might be simplest to do a small refactor to create a class level `handled errors` tuple, and then use that in each try/except block to handle.

## User-facing consequences
Very limited, has only been observed once in testing.

## Context
First reported here: https://github.com/learningequality/kolibri/pull/11765#issuecomment-1906241734

### PR DESCRIPTION
**Pull request by thesujai**
## Summary

The new approach consolidates handling of `sqlite3.OperationalError` and
`AssertionError` across multiple cache methods, also it makes the
internal implementation more cleaner without breaking anything
…

## References

Fixes #11767 
…

## Reviewer guidance

Confused about the naming, so i gave functions descriptive names

*please take your time to review it, no rush*
…

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by thesujai** [comment]
py 2.7 tests are failing because 0-args super() is supported from py 3.0
Maybe re-targetting to develop may work
or if that conflicts with the `milestones`, let me know so I can make change here to support py2.7 :)

File: kolibri/deployment/default/custom_django_cache.py
Comment (rtibbles) [inline kolibri/deployment/default/custom_django_cache.py]: I think you could simplify this all somewhat by having a single method like this:
```
def try_execute(self, method_name, error_return_value, *args, **kwargs):
    try:
        method = getattr(super(CustomDjangoCache, self), method_name)
        if method is None:
            raise ValueError("{method_name} is not a valid method".format(method_name=method_name))
        return method(*args, **kwargs)
    except self.ERRORS_TO_HANDLE:
        return error_return_value
```

That way you can guarantee you're using the Py2.7 compatible form of super, and you don't need to define 4 different functions that are doing the same thing.

If a function doesn't return a value explicitly in Python, it returns `None`, so you can pass `None` as the `error_return_value` in those cases.

**Review by rtibbles — CHANGES_REQUESTED** [review]
We do indeed want this to be Python 2.7 compatible, and most of the instances you have written are.

I have suggested a way to DRY this up slightly, which will also solve this, but you could equally just not call super in that way in the couple of instances you have.

File: kolibri/deployment/default/custom_django_cache.py
Comment (thesujai) [reply kolibri/deployment/default/custom_django_cache.py]: Thanks for the code snippet, indeed this thought crossed my mind, but i was confused about `error_return_value` having types. Now it makes more sense to me

**Comment by thesujai** [comment]
I've incorporated the suggested changes, let me know for additional adjustments or improvements at your convenience

File: kolibri/deployment/default/custom_django_cache.py
Comment (rtibbles) [inline kolibri/deployment/default/custom_django_cache.py]: Sorry, just one more thing - I think this probably does work to pass positional arguments as named keyword arguments, but I would find it less confusing if `method_name` and `error_return_value` were passed just as positional arguments in each instance here, so:
```
return self.try_execute(
    "add",
    False,
```

And so on.

Comment (thesujai) [reply kolibri/deployment/default/custom_django_cache.py]: sure, sorry for the trouble

**Review by rtibbles — COMMENTED** [review]
This is great, thanks @thesujai. I will hold off on final approval here, as we are only merging critical bug fixes right now into release-v0.16.x in anticipation of our 0.16.0 release.

We will merge once that release has been finalized!

**Review by rtibbles — APPROVED** [review]

**Pull request merged**
